python/revng/internal/api/rich_manager.py

```diff
@@ -0,0 +1,226 @@
+
+
+class StorageSaver(Thread):
+    def __init__(self, manager: Manager, synchronizer: Synchronizer, save_hooks: SaveHooks):
+        super().__init__()
+        self.manager = manager
+        self.synchronizer = synchronizer
+        self.save_hooks = save_hooks
+        self.stopping = False
+        self.shutdown_event = Event()
+
+        self.lock = Lock()
+        self.submit_event = Event()
+        self.dir_list: list[str] = []
+        self.credentials = ""
+
+    def run(self):
+        while True:
+            with self.lock:
+                if len(self.dir_list) == 0 and self.stopping:
+                    break
+
+            result = self.submit_event.wait(timeout=10)
+            if not result:
+                continue
+
+            with self.lock:
+                old_dir_list = self.dir_list.copy()
+                self.dir_list.clear()
+                self.submit_event.clear()
+
+            last_dir = Path(old_dir_list[-1])
+            old_directories = [Path(p) for p in reversed(old_dir_list[:-1])]
+
+            try:
+                self.synchronizer.save(last_dir, old_directories)
+            except self.synchronizer.save_exceptions() as e:
+                sys.stderr.write("Failed syncing resume directory\n")
+                traceback.print_exception(e, file=sys.stderr)
+                sys.stderr.flush()
+
+            for element in old_dir_list:
+                rmtree(element, False)
+
+            with self.lock:
+                credentials = self.credentials
+
+            for hook in self.save_hooks:
+                hook(self.manager, credentials)
+
+        self.shutdown_event.set()
+
+    def submit(self, dir_: str):
+        assert not self.stopping, "Cannot submit while stopping"
+        with self.lock:
+            self.dir_list.append(dir_)
+            self.submit_event.set()
+
+    def set_credentials(self, credentials: str):
+        with self.lock:
+            self.credentials = credentials
+            self.synchronizer.set_credentials(credentials)
+
+
+class ResettableTimer(Thread):
+    def __init__(self, interval: int, func: Callable[[], Any]):
+        super().__init__()
+        self.interval = interval
+        self.func = func
+        self.running = True
+
+        self.lock = Lock()
+        self.next_trigger_time = None
+
+    def run(self):
+        while self.running:
+            if self.next_trigger_time is not None and time.time() > self.next_trigger_time:
+                with self.lock:
+                    self.func()
+                    self.next_trigger_time = None
+            time.sleep(10)
+
+    def reset_countdown(self):
+        with self.lock:
+            self.next_trigger_time = None
+
+    def start_countdown(self):
+        with self.lock:
+            if self.next_trigger_time is None:
+                self.next_trigger_time = time.time() + self.interval
+
+
+class EventType(enum.StrEnum):
+    VOLATILE_CHANGE = enum.auto()
+    SAVE = enum.auto()
```

Comment:
I'm not sure we need this. Too much decoupling.

Can't you simply wrap certain functions with a function and certain other with another?

The event system seems overkill.

Reply:
> Can't you simply wrap certain functions with a function and certain other with another?

I'm not sure I follow, the objectives of the `RichManager` class are:
* To allow using a synchronizer if needed
* To allow triggering the event handler with some events, to trigger the autosave thread